src/rdkafka_broker.c

```diff
@@ -2539,12 +2540,13 @@ static void rd_kafka_broker_map_partitions (rd_kafka_broker_t *rkb) {
                         shptr_rd_kafka_toppar_t *s_rktp = rkt->rkt_p[i];
                         rd_kafka_toppar_t *rktp = rd_kafka_toppar_s2i(s_rktp);
 
-                        /* Only map unassigned partitions matching this broker*/
+                        /* Only map undelegated partitions matching this broker*/
                         rd_kafka_toppar_lock(rktp);
-                        if (rktp->rktp_leader_id == rkb->rkb_nodeid &&
-                            !(rktp->rktp_leader && rktp->rktp_next_leader)) {
-                                rd_kafka_toppar_leader_update(
-                                        rktp, rktp->rktp_leader_id, rkb);
+                        if (rktp->rktp_broker_id == rkb->rkb_nodeid &&
+                            !(rktp->rktp_broker && rktp->rktp_next_broker)) {
+                                rd_kafka_toppar_broker_update(
+                                        rktp, rktp->rktp_broker_id, rkb,
+                                        rd_true);
```

Comment:
right. i believe the behavior is correct as is because `rktp_leader_id` will never be different from `rktp_broker_id` if `rktp_broker` is `NULL`, however it's bad form to write this. i believe switching out `rktp_broker_id` for `rktp_leader_id` here is what we want.

generally, I found this all quite tricky to think through. I have a feeling a better structure is possible (but it's not immediately apparent what). Please check the interplay between `rktp_broker`, `rktp_leader_id`, and `rktp_broker_id` very carefully.